framework/web/UrlRule.php

```diff
@@ -256,6 +355,25 @@ public function parseRequest($manager, $request)
         }
         $matches = $this->substitutePlaceholderNames($matches);
 
+        // if routing succeeded (some rule matched), check whether pathInfo
+        //  was changed during canonization. If so, take the required action.
+        if ($pathInfoCanonized != $pathInfoOrig) {
+            switch ($this->canonize['action']) {
+                case 'redirect':
+                    $url = '/' . $pathInfoCanonized
+                        . (Yii::$app->request->queryString ? '?' . Yii::$app->request->queryString : '');
+                    die('redirect ' . $url);
+                    Yii::$app->getResponse()->redirect($url)->send();
+                    break;
```

Comment:
Redirection code injected to the point where rule already matched. Current yii2 logic (without canonizer feature) will just process this rule anyway - either action will be run, or NotFound exception will be thrown.

But I just thought - maybe we should return a route instead of explicit redirect? Something like

```
        'ruleConfig' => [
            'class' => 'yii\web\UrlRule',
            'canonize' => [
                'trailing-slash' => true,
                'collapse-slashes' => true,
                'action' => 'redirect',
                'redirect-route' => 'site/redirect'
            ]
        ],
```

It then would return ```[ 0 => 'site/redirect, 1 => ['redirect-url' => $pathInfoCanonized ]]```